tokio/src/time/sleep.rs

```diff
@@ -412,18 +448,34 @@ impl Sleep {
 
         #[cfg(any(not(tokio_unstable), not(feature = "tracing")))]
         let coop = ready!(crate::runtime::coop::poll_proceed(cx));
-
-        let result = me.entry.poll_elapsed(cx).map(move |r| {
-            coop.made_progress();
-            r
-        });
+        // Safety: we have just assigned it a value of `Some`.
+        let result = me
+            .entry
+            .as_pin_mut()
+            .unwrap()
+            .poll_elapsed(cx)
+            .map(move |r| {
+                coop.made_progress();
+                r
+            });
 
         #[cfg(all(tokio_unstable, feature = "tracing"))]
         return trace_poll_op!("poll_elapsed", result);
 
         #[cfg(any(not(tokio_unstable), not(feature = "tracing")))]
         return result;
     }
+
+    // This lazy initiation is for performance purposes,
+    // it can avoid the unnecessary creation and drop of `TimerEntry`.
+    fn lazy_init_timer_entry(self: Pin<&mut Self>, deadline: Instant) {
+        let mut me = self.project();
+        *me.deadline = deadline;
+        if me.entry.is_none() {
+            let entry = TimerEntry::new(me.handle, deadline);
+            me.entry.as_mut().set(Some(entry));
+        }
+    }
 }
```

Comment:
Drive-by review: could this be simplified with [`Option::get_or_insert_with`](https://doc.rust-lang.org/1.63.0/std/option/enum.Option.html#method.get_or_insert_with)?

Reply:
Thanks!